### ISSUE TITLE
Misc charts codebase improvements

### ISSUE THREAD

**Issue by bryevdv** [bokeh/bokeh#1773]
The `bokeh.charts` codebase could use some general improvements:

* [x] remove all `duplet`, `quartet`, `sextet`, `nonet`, etc in favor of readable `namedtuple`s (#1814)
* [x] `Step` should use `Line`, not `Segment`, because: (#1815)
  - otherwise impossible to add good hit-testing later
  - `Line` will miter the segments together in a more visually attractive way
* [x] remove all commented out code throughout
* [x] add `bokeh-plot` examples to docstrings (#1816)
* [x] split Builder class definitions into a `builders` subdirectory (#1817)
* [x] utilize `bokeh.properties` for private models, or make `Builders` be models (see below)  (#1819)
* [x] move `bokeh.charts` unit tests to `bokeh/charts/tests`  (#1818)
 - [x] separate test files for each module.

**Comment by bryevdv** [comment]
Ping @fpliger, @damianavila thoughts?

**Comment by bryevdv** [comment]
I also had an idea about how to simplify and unify code like this:
```
super(Chart, self).__init__(**kw)

self.__title = title
self.__xlabel = xlabel
self.__ylabel = ylabel
......
``` 
This is a perfect opportunity to re-use our properties and models system to make things more self-documenting, as well as get a bunch of validation for free. Basically, use `PlotObject` as "private mode" and get all the good stuff that models provide for free. I imagine this:
```

class _ChartArgs(PlotObject):
        title = Str
        xlabel = Str
        ylabel = Str
        legend = Bool
        xscale = Float
        yscale = Float
        width = Int
        height = Int
        tools = List(Str)
        filename = Str
        server = Str
        notebook = Bool
        facet = FacetEnum
        palette = Either(Palette, Str)
        xgrid = True
        ygrid = True
```
Then we can gut the entire `Chart` initializer:
```
def __init__(self, foo, session=None, **kws):
    
    # Use foo, session, etc. If need be, pick known values out of **kws (remove them)

    # Anything that does not belong or is the wrong type gets caught here!
    self._args = _ChartArgs(**kw)
```
I think methods like `check_attrs` could be automated away if we did this too. 

This is just a bare outline and there are certainly lots of details but I think this would be a great enhancement to overall reliability and maintainability.

**Comment by bryevdv** [comment]
Alternatively, we could just make the `Builders` be models (subclass of `PlotObject`). In fact, that's probably the best plan. Any `Builder` would get automatic validation for all its properties, and easy auto-documentation.

**Comment by fpliger** [comment]
I like almost all proposals. The only one I really don't get is:

>Alternatively, we could just make the Builders be models (subclass of PlotObject). In fact, that's probably the best plan. Any Builder would get automatic validation for all its properties, and easy auto-documentation.

Builders are already quite lean (only handle the values creation/computation, palette, legend, ranges... and a few other internal things). I see big benefits regards `Chart` and `ChartArgs` (really painful to keep and maintain attrs (with like `check_attrs`). IMHO optimal scenario would be if builders could just take care of:

- compute data (and build source as a consequence)
- draw the renderers
- make legends
- (maybe ranges?)

**Comment by damianavila** [comment]
>remove all duplet, quartet, sextet, nonet, etc in favor of readable namedtuples

I agree... looking at those now seems very confusing (probably very confusing from the beginning

>Step should use Line, not Segment, because:
otherwise impossible to add good hit-testing later
Line will miter the segments together in a more visually attractive way

Sound good...

>remove all commented out code throughout

:+1: 

>add bokeh-plot examples to docstrings

:+1: 

>split Builder class definitions into a builders subdirectory

you mean a builders subdirectory containing each builder and a charts.py (or something else) containing all the functions to create the actual charts? 

>utilize bokeh.properties for private models, or make Builders be models (see below)

:+1: , really nice idea...

**Comment by damianavila** [comment]
>Alternatively, we could just make the Builders be models (subclass of PlotObject). In fact, that's probably the best plan. Any Builder would get automatic validation for all its properties, and easy auto-documentation.

Yep, probably the best idea...

**Comment by bryevdv** [comment]
@damian something like this:
```
bokeh/charts/
    __init__.py
    builders/
        heatmap_builder.py
        horizon_builder.py
        line_builder.py
        scatter_builder.py
        timeseries_builder.py
    builder.py
    chart.py
    data_adapter.py
    utils.py
```
The builder subclasses would be explicitly named `LineBuilder`, `ScatterBuilder`, etc.

Then the only question is: where do the user-facing functions, `TimeSeries`, etc go? A few options:
* in individual `line.py` modules, so have: `bokeh.charts.line.Line`
* explicitly defined in `__init__.py` so have: `bokeh.charts.Line`
* imported into `__init__.py` from `line.py`, so have: both of the above
* auto-generated somehow in `__init__.py` so have:  `bokeh.charts.Line`

The code for the user functions is almost all boilerplate, my personal preference would be for the last one. I would imagine some kind of extra information being stored in the builder modules, then the `__init__.py` would scan the builders sub packages and create and expose a user function for every builder.

**Comment by damianavila** [comment]
3rd or 4rd options... the 4th one deserves more time, but it would be nice to have it... in the meantime the 3rd is probably enough... thx for the clarification...

**Comment by bryevdv** [comment]
I agree 3rd is probably god to start, then I would actually recommend module names like `_line.py` (so then: `_line.Line` so that users don't get used to an idea that they should be using them.

**Comment by damianavila** [comment]
> I agree 3rd is probably god to start, then I would actually recommend module names like _line.py (so then: _line.Line) so that users don't get used to an idea that they should be using them.

:+1:

**Comment by bryevdv** [comment]
> Builders are already quite lean (only handle the values creation/computation, palette, legend, ranges... and a few other internal things). I see big benefits regards Chart and ChartArgs (really painful to keep and maintain attrs (with like check_attrs). IMHO optimal scenario would be if builders could just take care of

Well, I definitely like the idea of explicitly laying out what high-level tasks a builder is responsible for. If all builders follow a prescribed pattern then it will be much easier to document and explain to others how to make new ones. I agree somewhat about builders being lean, except noting that "compute data" might be a very heavy and involved operation in some cases. That said, I think making builders also be models helps make them *more* lean, not less. Builders will inevitably have per-builder properties (think of statistical computation parameters for histograms, etc., or number of rings in a donut chart). Using properties  bus things like validation and crystal clear specification for these things. 

Basically my experience with `autodoc` taught me two things:
* plug-in frameworks can be amazing for reducing the work needed to extend libraries, *BUT*
* they are worthless unless the expectations and patterns are documented (people have to know what parts to plug in)

This and the last comment is about pushing `bokeh.charts` in this "pluggable" direction, isolating all the parts that are truly required to extend `bokeh.charts` with a new chart type, and putting them in one place (e.g. builders subdir), documenting the pattern (as @fpliger described), and saving people from the burden of having to write lots of boilerplate (using our great properties machinery)

**Comment by fpliger** [comment]
Thanks for the details @bryevdv. With that extra context I do agree with you. 

> Then the only question is: where do the user-facing functions, TimeSeries, etc go?

I really like the direction that the discussion is taking. The pattern is really nice. With that in mind I agree with the line that 3rd option is just the easier one and works ok for now. That said I think the 4th option is the right direction for a really flexible, extendable and pluggable design.
